test/cloudtest/pkg/providers/packet/packet_provider.go

```diff
@@ -0,0 +1,584 @@
+	}
+
+	return facilitiesList, nil
+}
+
+func (pi *packetInstance) Destroy(timeout time.Duration) error {
+	logrus.Infof("Destroying cluster  %s", pi.id)
+	if (pi.client != nil) {
+		response, err := pi.client.SSHKeys.Delete(pi.sshKey.ID)
+		pi.manager.AddLog(pi.id, "delete-sshkey", fmt.Sprintf("%v\n%v\n%v", pi.sshKey, response, err))
+		for key, device := range pi.devices {
+			response, err := pi.client.Devices.Delete(device.ID)
+			pi.manager.AddLog(pi.id, fmt.Sprintf("delete-device-%s", key), fmt.Sprintf("%v\n%v", response, err))
+		}
+	}
+	return nil
+}
+
+func (pi *packetInstance) GetRoot() string {
+	return pi.root
+}
+
+func (pi *packetInstance) doDestroy(writer io.StringWriter, timeout time.Duration, err error) {
+	_, _ = writer.WriteString(fmt.Sprintf("Error during k8s API initialisation %v", err))
+	_, _ = writer.WriteString(fmt.Sprintf("Trying to destroy cluster"))
+	// In case we failed to start and create cluster utils.
+	err2 := pi.Destroy(timeout)
+	if err2 != nil {
+		_, _ = writer.WriteString(fmt.Sprintf("Error during destroy of cluster %v", err2))
+	}
+}
+
+func (pi *packetInstance) doInstall(context context.Context) error {
+	pi.provider.Lock()
+	defer pi.provider.Unlock()
+	if pi.installScript != nil && !pi.provider.installDone[pi.config.Name] {
+		pi.provider.installDone[pi.config.Name] = true
+		return pi.shellInterface.RunCmd(context, "install", pi.installScript, nil)
+	}
+	return nil
+}
+
+func (pi *packetInstance) updateProject() error {
+	ps, response, err := pi.client.Projects.List(nil)
+
+	out := strings.Builder{}
+	_, _ = out.WriteString(fmt.Sprintf("%v\n%v\n", response, err))
+
+	if err != nil {
+		logrus.Errorf("Failed to list Packet projects")
+	}
+
+	for i := 0; i < len(ps); i++ {
+		p := &ps[i]
+		_, _ = out.WriteString(fmt.Sprintf("Project: %v\n %v", p.Name, p))
+		if p.ID == pi.projectID {
+			pp := ps[i]
+			pi.project = &pp
+		}
+	}
+
+	pi.manager.AddLog(pi.id, "list-projects", out.String())
+
+	if pi.project == nil {
+		err := fmt.Errorf("%s - specified project are not found on Packet %v", pi.id, pi.projectID)
+		logrus.Errorf(err.Error())
+		return err
+	}
+	return nil
+}
+
+func (pi *packetInstance) createKey(keyFile string) ([]string, error) {
+	pi.keyID = "dev-ci-cloud-" + pi.genID
+
+	out := strings.Builder{}
+	keyFileContent, err := utils.ReadFile(keyFile)
+	if err != nil {
+		_,_ = out.WriteString(fmt.Sprintf("Failed to read key file %s", keyFile))
+		pi.manager.AddLog(pi.id, "create-key", out.String())
+		logrus.Errorf("Failed to read file %v %v", keyFile, err)
+		return nil, err
+	}
+
+	_,_ = out.WriteString(fmt.Sprintf("Key file %s readed ok", keyFile))
+
+	keyRequest := &packngo.SSHKeyCreateRequest{
+		ProjectID: pi.project.ID,
+		Label:     pi.keyID,
+		Key:       strings.Join(keyFileContent, "\n"),
+	}
+	sshKey, response, err := pi.client.SSHKeys.Create(keyRequest)
+
+	createMsg := fmt.Sprintf("Create key %v %v %v", sshKey, response.String(), err)
+	_,_ = out.WriteString(createMsg)
```

Comment:
I think it is rather a sensitive data, could we avoid printing it?

Reply:
This is a public key, no worries :-)